src/ModelObjects/component.rs

```diff
@@ -659,6 +659,69 @@ pub enum SyncType {
     Output,
 }
 
+pub type DecoratedLocationTuple<'a> = Vec<DecoratedLocation<'a>>;
+
+//Represents a single transition from taking edges in multiple components
+#[derive(Debug, Clone)]
+pub struct Transition<'a> {
+    pub edges: Vec<(&'a Component, &'a Edge, usize)>,
+}
+impl<'a> Transition<'a> {
+    pub fn combinations(left: &mut Vec<Self>, right: &mut Vec<Self>) -> Vec<Self> {
+        let mut out = vec![];
+        for l in left {
+            for r in &*right {
+                let temp: Vec<(&'a Component, &'a Edge, usize)> = l
+                    .edges
+                    .iter()
+                    .cloned()
+                    .chain(r.edges.iter().cloned())
+                    .collect();
+                out.push(Transition { edges: temp });
+            }
+        }
+
+        out
+    }
+
+    pub fn apply_updates(&self, locations: &mut DecoratedLocationTuple, zone: &mut Zone) {
+        for (comp, edge, index) in &self.edges {
+            edge.apply_update(&mut locations[*index], zone);
+        }
+    }
+
+    pub fn apply_guards(&self, locations: &DecoratedLocationTuple, zone: &mut Zone) -> bool {
+        let mut success = true;
+        for (comp, edge, index) in &self.edges {
+            success = success && edge.apply_guard(&locations[*index], zone);
+        }
+        success
+    }
+
+    pub fn apply_guards_after_invariants(
+        &self,
+        locations: &DecoratedLocationTuple,
+        zone: &mut Zone,
+    ) -> bool {
+        let mut success = true;
+        for (comp, edge, index) in &self.edges {
+            success = success
+                && locations[*index].apply_invariant(zone)
+                && edge.apply_guard(&locations[*index], zone);
+        }
+        success
+    }
```

Comment:
I think we might be applying the wrong invariant here, to my understanding we want to apply the invariant for the transitions target locations. But `locations[*index]` gives the source locations.

Although its implemented in the same way as before.